strawberryfields/circuitspecs/x.py

```diff
@@ -0,0 +1,256 @@
+
+        # Use the GaussianUnitary compiler to compute the symplectic
+        # matrix representing the Gaussian operations.
+        # Note that the Gaussian unitary compiler does not accept measurements,
+        # so we append the measurement separately.
+        meas_seq = [seq[-1]]
+        seq = GaussianUnitary().compile(seq[:-1], registers) + meas_seq
+
+        # determine the modes that are acted on by the symplectic transformation
+        used_modes = [x.ind for x in seq[0].reg]
+
+        # extract the compiled symplectic matrix
+        S = seq[0].op.p[0]
+
+        if len(used_modes) != n_modes:
+            # The symplectic transformation acts on a subset of
+            # the programs registers. We must expand the symplectic
+            # matrix to one that acts on all registers.
+            # simply extract the computed symplectic matrix
+            S = expand(seq[0].op.p[0], used_modes, n_modes)
+
+        half_n_modes = n_modes // 2
+
+        # Construct the covariance matrix of the state.
+        # Note that hbar is a global variable that is set by the user
+        cov = (sf.hbar / 2) * S @ S.T
+
+        # Construct the A matrix
+        A = Amat(cov, hbar=sf.hbar)
+
+        # Construct the adjacency matrix represented by the A matrix.
+        # This must be an weighted, undirected bipartite graph. That is,
+        # B00 = B11 = 0 (no edges between the two vertex sets 0 and 1),
+        # and B01 == B10.T (undirected edges between the two vertex sets).
+        B = A[:n_modes, :n_modes]
+        B00 = B[:half_n_modes, :half_n_modes]
+        B01 = B[:half_n_modes, half_n_modes:]
+        B10 = B[half_n_modes:, :half_n_modes]
+        B11 = B[half_n_modes:, half_n_modes:]
+
+        # Perform unitary validation to ensure that the
+        # applied unitary is valid.
+
+        if not np.allclose(B00, 0) or not np.allclose(B11, 0):
+            # Not a bipartite graph
+            raise CircuitError(
+                "The applied unitary cannot mix between the modes {}-{} and modes {}-{}.".format(
+                    0, half_n_modes - 1, half_n_modes, n_modes - 1
+                )
+            )
+
+        if not np.allclose(B01, B10):
+            # Not a symmetric bipartite graph
+            raise CircuitError(
+                "The applied unitary on modes {}-{} must be identical to the applied unitary on modes {}-{}.".format(
+                    0, half_n_modes - 1, half_n_modes, n_modes - 1
+                )
+            )
+
+        # Now that the unitary has been validated, perform the Takagi decomposition
+        # to determine the constituent two-mode squeezing and interferometer
+        # parameters.
+        sqs, U = takagi(B01)
+        sqs = np.arctanh(sqs)
+
+        # ensure provided S2gates all have the allowed squeezing values
+        if not all(s in self.allowed_sq_ranges for s in sqs):
+            wrong_sq_values = [s for s in sqs if s not in self.allowed_sq_ranges]
```

Comment:
Minor optimization: only need to perform the range check once.
```suggestion
        wrong_sq_values = [s for s in sqs if s not in self.allowed_sq_ranges]
        if wrong_sq_values :
```